backend/cmd/worker/sample_validation.go

```diff
@@ -0,0 +1,260 @@
+			}
+		}
+		return nil
+	})
+}
+
+func insertSampleRecords(records []SampleValidationRecord, repo repository.SamplesDAO) error {
+	for _, record := range records {
+		if !record.Skipped {
+			sample := types.Sample{
+				// TODO: Make sure all fields are correctly mapped
+				TypeCode:                record.Sample.TypeCode,
+				SubmitterSampleId:       record.Sample.SubmitterSampleId,
+				TissueType:              record.Sample.TissueSite,
+				HistologyTypeCode:       record.Sample.HistologyCode,
+				SubmitterOrganizationId: record.OrganizationId,
+			}
+			err := repo.CreateSample(&sample)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func validateSamplesBatch(samples []types.SampleBatch, repoOrganization repository.OrganizationDAO, repoPatient repository.PatientsDAO, repoSample repository.SamplesDAO) ([]SampleValidationRecord, error) {
+	var records []SampleValidationRecord
+
+	for index, sample := range samples {
+		record := &SampleValidationRecord{
+			Sample: sample,
+			BaseValidationRecord: BaseValidationRecord{
+				Index: index,
+			},
+		}
+
+		// 1. Validate fields
+		record.validateFieldLength("submitter_parent_sample_id", sample.SubmitterParentSampleId)
+		record.validateFieldLength("tissue_site", sample.TissueSite)
+
+		// 2. Validate patient
+		patient, patientErr := repoPatient.GetPatientByOrgCodeAndOrgPatientId(sample.PatientOrganizationCode, sample.SubmitterPatientId)
+		if patientErr != nil {
+			return nil, fmt.Errorf("error getting existing patient: %v", patientErr)
+		}
+		record.validatePatient(patient)
+
+		// 3. Validate organization
+		organization, orgErr := repoOrganization.GetOrganizationByCode(sample.SampleOrganizationCode)
+		if orgErr != nil {
+			return nil, fmt.Errorf("error getting existing sample organization: %v", orgErr)
+		}
+		record.validateOrganization(organization)
+
+		// 4. Validate if sample exists in DB
+		if organization != nil {
+			existingSample, sampleErr := repoSample.GetSampleBySubmitterSampleId(organization.ID, sample.SubmitterSampleId)
+			if sampleErr != nil {
+				return nil, fmt.Errorf("error getting existing sample: %v", sampleErr)
+			} else if existingSample != nil {
+				// 5. If exists, check if all fields are identical, and add error messages
+				record.validateExistingSampleInDb(existingSample)
+			}
+
+			// 6. Validate parent sample in DB if provided
+			if sample.SubmitterParentSampleId != "" {
+				existingParentSample, parentSampleErr := repoSample.GetSampleBySubmitterSampleId(organization.ID, sample.SubmitterParentSampleId)
+				if parentSampleErr != nil {
+					return nil, fmt.Errorf("error getting existing parent sample: %v", parentSampleErr)
+				}
+				record.validateExistingParentSampleInDb(existingParentSample)
+
+				if existingParentSample == nil {
+					// 7. If parent sample does not exist in DB, check if it exists in the current batch
+					parentSampleInBatch := slices.ContainsFunc(samples, func(s types.SampleBatch) bool {
```

Comment:
suggestion : build a map of SampleKey{sample_organization_code, sample_id} and then lookup in the map for searching a parent. It should reduce complexity from O(n2) to O(n)